### DISCUSSION TITLE
How can I get current session object in mapper event or get objects to commit in the session event

### DISCUSSION THREAD

**Discussion by PaleNeutron** [sqlalchemy/sqlalchemy#11679] [Usage Questions] [answered]
I want to create some log rows when an ORM table changes. So I use `@event.listens_for(MyModel, "before_update")` to do that and create session from the connection using:

```
with Session(connection) as session:
    ... # create log obj
    session.add(log)
    session.commit()
```

But this event is called each object update and will commit thousand times.

How to improve this logic? It seems I can not get `target` in `before_commit` event.

**Comment by zzzeek** [comment, marked as answer]
well technically if you are using the connection passed to the before_update event and sending that to a new Session, it isn't actually doing a DB COMMIT because it's joining onto the connection that's already in progress.   so you are mostly using a lot of Python CPU to build up a new session and all that.

the most efficient way here from a Python perspective is to just use an insert() statement on the connection:

```py
connection.execute(
    my_log_table.insert(), {"message": "did a thing"}
)
```

if you were interested in batching those INSERTs into one big insert, then I'd likely use the `after_flush` method.  you would iterate through the `session.dirty` collection and locate all the objects you want to set up for logging (use isinstance() to see if its the class(es) you care about), then build up a list of log messages on a per-table basis.
